packages/next-on-pages/src/buildApplication/packageManagerUtils.ts

```diff
@@ -53,91 +77,204 @@ export async function getCurrentPackageManager(): Promise<PackageManager> {
+			packageVersion = pm.startsWith('yarn')
 				? packageInfo?.children?.Version
-				: packageInfo?.dependencies[packageName]?.version;
+				: packageInfo?.dependencies?.[packageName]?.version ??
+				  packageInfo?.devDependencies?.[packageName]?.version;
+		}
+
 		return packageVersion ?? null;
 	} catch {
 		return null;
 	}
 }
 
-export function getBinaryVersion(binaryName: PackageManager): string | null {
-	const commandArgs = ['--version'];
+/**
+ * Gets the version of a binary installed on the machine.
+ *
+ * @param binaryName Name of the binary to get the version of.
+ * @returns Version of the binary, or `null` if the binary is not installed.
+ */
+export async function getBinaryVersion(
+	binaryName: string,
+): Promise<string | null> {
+	const versionFromProcess = process.versions[binaryName];
+	if (versionFromProcess) return versionFromProcess;
+
 	try {
-		return execFileSync(getPackageManagerSpawnCommand(binaryName), commandArgs)
+		const cmd = isWindows() ? '.cmd' : '';
+		return execFileSync(`${binaryName}${cmd}`, ['--version'])
 			.toString()
-			.trim();
+			.trim()
+			.replace(/^v/, '');
 	} catch {
 		return null;
 	}
 }
+
+/**
+ * Waits for a spawned process to close.
+ *
+ * @param spawnedProcess Spawned process to wait for.
+ * @returns Promise that resolves when the process closes with code 0, or rejects when the process
+ * closes with a non-zero code.
+ */
+export async function waitForProcessToClose(
+	spawnedProcess: ReturnType<typeof spawn>,
+): Promise<void> {
+	return new Promise((resolve, reject) => {
+		spawnedProcess.on('close', code => {
+			if (code === 0) {
+				resolve();
+			} else {
+				reject();
+			}
+		});
+	});
+}
+
+/**
+ * Gets the executable string used for the package manager.
+ *
+ * @param pm Package manager name.
+ * @param packageName Package name.
+ * @returns The executable string used for the package manager.
+ */
+export async function getExecStr(
+	pm: PackageManager,
+	packageName?: string,
+): Promise<string> {
+	const {
+		baseCmd,
+		execCmd,
+		dlxOrExec,
+		execArgs = [],
+	} = await getPackageManagerInfo(pm);
+
+	let dlxArgs: string[] = [];
+	if (packageName && dlxOrExec) {
+		const vercelPackageIsInstalled = await getPackageVersion(packageName, pm);
+		dlxArgs = dlxOrExec(!vercelPackageIsInstalled);
+	}
+
+	const args = [...dlxArgs, ...execArgs];
+
+	return `${execCmd ?? baseCmd} ${args.join(' ') ?? ''}`.trim();
+}
+
+/**
+ * Checks whether the current platform is Windows.
+ *
+ * @returns Whether the current platform is Windows.
+ */
+function isWindows(): boolean {
+	return process.platform === 'win32';
+}
+
+export type PackageManager =
+	| 'bun'
+	| 'pnpm'
+	| 'yarn (berry)'
+	| 'yarn (classic)'
+	| 'yarn'
+	| 'npm';
+
+export type PackageManagerInfo = {
+	pm: PackageManager;
+	baseCmd: string;
+	execCmd?: string;
+	execArgs?: string[];
+	dlxOrExec?: (useDlx: boolean) => string[];
+	infoArgs: string[];
+	infoRegex?: (name: string) => RegExp;
+};
```

Comment:
I think that here we could have more strict typing if we wanted

It could make things slightly clearer/more robust, but I don't think it's too important so it's just a thought, I'm happy to go with `string`s if you prefer them

```suggestion
type PlainPackageManager = Exclude<
	PackageManager,
	'yarn (berry)' | 'yarn (classic)'
>;
type PackageManagerBaseCmd = `${PlainPackageManager}${'' | '.cmd'}`;

export type PackageManagerInfo = {
	pm: PackageManager;
	baseCmd: PackageManagerBaseCmd;
	execCmd?: `npx${'' | '.cmd'}`;
	execArgs?: string[];
	dlxOrExec?: (useDlx: boolean) => ('dlx' | 'exec')[];
	infoArgs: string[];
	infoRegex?: (name: string) => RegExp;
};
```